### ISSUE TITLE
Support for Multiple Data Stores

### ISSUE THREAD

**Issue by darylrobbins** [NREL/api-umbrella#85]
I am looking at porting api-umbrella to use Amazon DynamoDB instead of MongoDB. The current design is using Mongo-specific ORM libraries in the gatekeeper and web interface. Any thoughts on how we could abstract away from any specific storage provider, so I can contributed my code back to the project? Thanks!

**Comment by GUI** [comment]
Unfortunately, I think this will be a bit tricky as things exist right now. While it might be possible to abstract things at the ORM-level ([jugglingdb](https://github.com/1602/jugglingdb) might be a candidate for the NodeJS side of things, but I don't know of any equivalent abstraction for the Ruby world), I think we are somewhat reliant on a couple of MongoDB characteristics that don't exist in DynamoDB. For example, we use MongoDB's [TTL capabilities](http://docs.mongodb.org/manual/tutorial/expire-data/) to distribute rate limit information among a multi-node cluster. I'm not super familiar with DynamoDB, but a quick search seems to indicate it doesn't support this specific feature. It's certainly possible to address this kind of thing in other ways, it just becomes trickier, since swapping MongoDB for DynamoDB isn't a one-to-one mapping.

But we'd certainly be open to these kind of abstractions. I don't necessarily want to be tied to MongoDB, it just happens to be the database we're currently using for this project. So while we have become somewhat tied to it, since that's simply what we've been using, I don't think there's a ton of MongoDB specific functionality we require.

May I ask what's driving this, though? Is it the operational simplicity of DynamoDB, or something else? If it's that simplicity you get with running DynamoDB, would some MongoDB-as-a-service operating on AWS work for you? For example, something like [MongoLab](https://mongolab.com), [ObjectRocket](https://objectrocket.com), or [Compose.io](https://www.compose.io)?

**Comment by darylrobbins** [comment]
Yes, I stumbled upon jugglingdb as well. It looks like their DynamoDB adapter is still relatively immature.

You're right. This does sound like it could be quite tricky to accomplish. DynamoDB does not have any similar TTL capabilities. To expire the data in a systematic way like with the TTL feature, you'd either need a background janitor process and/or some strategy for bucketing your data into different tables based on time. It could get quite messy, quite fast.

I think a bit part of the problem is that ORM layers for NoSQL data stores are still pretty immature, not to mention that each such database works a bit differently and offers different capabilities. There is no mature ActiveRecord-type abstraction in the NoSQL world.

It almost seems ElasticSearch would be a better fit than DynamoDB in this case. It's got the TTL functionality and is essentially a document store with advanced text search capabilities.

Yes, our driver is definitely operational simplicity, but it's two-fold: there's the management side as you mentioned but then there'a also the number of technologies in our landscape side of things to look at. One big reason that I like DynamoDB (and why we're already using it) is that it's a hybrid key-value/document store. It may not have all the features of MongoDB and Cassandra, but it offers an excellent compromise, which meets our needs for both storage paradigms.

So, ideally, even if we didn't have to manage it ourselves, I'm looking to avoid introducing yet another storage technology into our stack. We already have DynamoDB, PostgreSQL, Microsoft SQL Server, ElasticSearch, and Redis. :)

**Comment by darylrobbins** [comment]
After considering it further, I think it's likely worth adding another storage technology in our stack to avoid going down the refactoring rabbit hole.

**Comment by darylrobbins** [comment]
You've got a great project here! I think we're going to invest our effort in improving the web interface instead and then we can contribute back our changes to the project.
